Initial Arch Documentation (#3051)
microsoft/msquic: docs/Architecture.md

Architecture
======

MsQuic has primarily two high-level layers: QUIC and the platform. The platform layer includes
abstractions for TLS, UDP and OS stuff (like threads and locks). The QUIC layer is platform
independent logic, built on the platform abstraction layers.

# Component Breakdown

## Platform Abstraction Layer

The platform abstraction layer (or PAL) supports the following platforms:

- Windows (User and Kernel)
- Linux
- FreeBSD
- macOS

### Basic OS Interfaces

> The most important and fundamental part of the PAL is the (mostly inline) definitions of may of the basic constructs exposed. Things like locks, event, interlocked operations, threads, time and more are all exposed. As mentioned above, most of these definitions are contained inline in a set of C headers, using `#define` macros. The goal here is for there to be little to no extra overhead from this layer; simply a minimal abstraction for the various interfaces.

Reviewer: ```suggestion
The most fundamental part of the PAL is definitions of basic constructs like locks, events, interlocked operations, threads, and time. Most of these definitions are implemented inline in C macros. The goal is to minimize overhead and simply be a minimal abstraction for various interfaces.
```

> ### Data Path

Reviewer: ```suggestion
### Datapath
```

> Another large piece of the PAL is the abstraction of UDP and TCP sockets, named the "data path" layer. This layer is designed to be highly parallelized and as performant as possible on the various platforms supported.

Reviewer: ```suggestion
Another large piece of the PAL is the abstraction of UDP and TCP sockets, named the "datapath" layer. This layer is designed to be highly parallelized and performant.
```

It uses asynchronous callbacks driven by the current execution model (see [below](#execution-model)) threads.

Currently, it also has preview support of XDP on Windows.

### Crypto

The PAL exposes various cryptographic interfaces as well, such has encryption/decryption and cryptographic hashing. This layer currently supports BCrypt and OpenSSL variants on Windows, and OpenSSL on Linux.

### TLS

Related to Crypto, the TLS layer abstracts the necessary TLS functionality required by the QUIC protocol. This layer currently supports Schannel and OpenSSL variants on Windows, and OpenSSL on Linux.

### Platform Initialization

> The platform layer isn't completely stateless, and requires some setup and cleanup. This are split into two separate layers:

Reviewer: ```suggestion
The platform layer isn't completely stateless, and requires some setup and cleanup. This is split into two separate layers:
```
